src/Microsoft.DotNet.ProjectModel/Files/IncludeContext.cs

```diff
@@ -0,0 +1,175 @@
+            return other != null &&
+                SourceBasePath == other.SourceBasePath &&
+                Option == other.Option &&
+                EnumerableEquals(IncludePatterns, other.IncludePatterns) &&
+                EnumerableEquals(ExcludePatterns, other.ExcludePatterns) &&
+                EnumerableEquals(IncludeFiles, other.IncludeFiles) &&
+                EnumerableEquals(ExcludeFiles, other.ExcludeFiles) &&
+                EnumerableEquals(BuiltInsInclude, other.BuiltInsInclude) &&
+                EnumerableEquals(BuiltInsExclude, other.BuiltInsExclude) &&
+                EnumerableEquals(Mappings, other.Mappings);
+        }
+
+        public override int GetHashCode()
+        {
+            return base.GetHashCode();
+        }
+
+        private static bool EnumerableEquals<T>(IEnumerable<T> left, IEnumerable<T> right)
+            => Enumerable.SequenceEqual(left ?? EmptyArray<T>.Value, right ?? EmptyArray<T>.Value);
+
+        private static string[] ExtractValues(JToken token)
+        {
+            if (token != null)
+            {
+                if (token.Type == JTokenType.String)
+                {
+                    return new string[] { token.Value<string>() };
+                }
+
+                else if (token.Type == JTokenType.Array)
+                {
+                    return token.Values<string>().ToArray();
+                }
+            }
+
+            return new string[0];
+        }
+
+        private static List<string> CreateCollection(string projectDirectory, string propertyName, IEnumerable<string> patternsStrings, bool literalPath)
+        {
+            var patterns = patternsStrings.SelectMany(patternsString => GetSourcesSplit(patternsString))
+                                          .Select(patternString => patternString.Replace('/', Path.DirectorySeparatorChar).Replace('\\', Path.DirectorySeparatorChar));
+
+            foreach (var pattern in patterns)
+            {
+                if (Path.IsPathRooted(pattern))
+                {
+                    throw new InvalidOperationException($"The '{propertyName}' property cannot be a rooted path.");
+                }
+
+                if (literalPath && pattern.Contains('*'))
+                {
+                    throw new InvalidOperationException($"The '{propertyName}' property cannot contain wildcard characters.");
+                }
+            }
+
+            return new List<string>(patterns.Select(pattern => FolderToPattern(pattern, projectDirectory)));
+        }
+
+        private static IEnumerable<string> GetSourcesSplit(string sourceDescription)
+        {
+            if (string.IsNullOrEmpty(sourceDescription))
+            {
+                return Enumerable.Empty<string>();
+            }
+
+            return sourceDescription.Split(PatternSeparator, StringSplitOptions.RemoveEmptyEntries);
+        }
+
+        private static string FolderToPattern(string candidate, string projectDir)
```

Comment:
this method looks familiar, is it copied from some old logic? if so can we reuse it instead of copying?